File: day01/cpp/day01.cpp
#include <cassert>
#include <cmath>
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

#include <gmpxx.h>


mpz_class part_one(std::vector<mpz_class> input) {
	mpz_class total = 0;
	for (auto mass : input)
		total += mass/3 - 2;

	return total;
}

mpz_class part_two(std::vector<mpz_class> input) {
	mpz_class total = 0;
	for (auto mass : input) {
		while ((mass = mass/3 - 2) > 0)
			total += mass;
	}
	return total;
}

int main() {
	std::ifstream raw_input("input");
	auto input = std::vector<mpz_class>();

	std::string line;
	while (std::getline(raw_input, line)) {
		mpz_class tmp;
		tmp.set_str(line, 10);
		input.push_back(tmp);
	}

	//std::cout << part_one(input) << std::endl;
	std::cout << part_two(input) << std::endl;
}

File: day01/day01.py
from aocd.models import Puzzle
import cProfile
from collections import defaultdict, namedtuple, Counter, deque
from itertools import permutations, combinations, chain
import re
import math


def part_one(_input):
    total = 0
    for mass in _input:
        total += mass//3 - 2
    return total


def part_two(_input):
    total = 0
    for mass in _input:
        while (mass := mass//3 - 2) > 0:
            total += mass
    return total


if __name__ == '__main__':
    puzzle = Puzzle(year=2019, day=1)
    _input = puzzle.input_data.split('\n')
    #_input = open('bigboy').readlines()

    _input = [int(line) for line in _input]

    print(part_one(_input))
    print(part_two(_input))

    #cProfile.run('print(part_one(_input))')
    #cProfile.run('print(part_two(_input))')
